deploy: enable HonorPVReclaimPolicy feature gate

yati1998 opened the pull request:
This commit enables the HonorPVReclaimPolicy feature
gate.

fixes: #3230

Signed-off-by: Yati Padia <ypadia@redhat.com>

nixpanic committed:
nfs: make DeleteVolume (more) idempotent

Signed-off-by: Niels de Vos <ndevos@redhat.com>

yati1998 committed:
deploy: enable HonorPVReclaimPolicy feature gate

This commit enables the HonorPVReclaimPolicy feature
gate.

fixes: #3230

Signed-off-by: Yati Padia <ypadia@redhat.com>

nixpanic commented:
@mergifyio rebase

nixpanic commented:
https://github.com/nixpanic/ceph-csi/tree/nfs/provisioner/DeleteVolume might have a fix for this in case `DeleteVolume` fails with errors like
```
GRPC error: rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory
```

nixpanic commented:
It seems some RBAC changes are needed for the NFS-provisioner:
```
I0720 12:02:56.874062       1 controller.go:1471] delete "pvc-2355cacb-b2da-4e6a-a91c-80022caa4daf": started
I0720 12:02:56.878091       1 connection.go:183] GRPC call: /csi.v1.Controller/DeleteVolume
I0720 12:02:56.878327       1 connection.go:184] GRPC request: {"secrets":"***stripped***","volume_id":"0001-0024-6be59c91-cd16-4814-811e-300690c167f3-0000000000000001-aed2ab46-0823-11ed-8c24-5e0233e9e911"}
I0720 12:02:56.997187       1 connection.go:186] GRPC response: {}
I0720 12:02:56.997283       1 connection.go:187] GRPC error: <nil>
I0720 12:02:56.997330       1 controller.go:1486] delete "pvc-2355cacb-b2da-4e6a-a91c-80022caa4daf": volume deleted
I0720 12:02:57.003979       1 controller.go:1523] delete "pvc-2355cacb-b2da-4e6a-a91c-80022caa4daf": failed to remove finalizer for persistentvolume: persistentvolumes "pvc-2355cacb-b2da-4e6a-a91c-80022caa4daf" is forbidden: User "system:serviceaccount:cephcsi-e2e-f0cfe643:nfs-csi-provisioner" cannot update resource "persistentvolumes" in API group "" at the cluster scope
W0720 12:02:57.004177       1 controller.go:989] Retrying syncing volume "pvc-2355cacb-b2da-4e6a-a91c-80022caa4daf", failure 0
E0720 12:02:57.004356       1 controller.go:1007] error syncing volume "pvc-2355cacb-b2da-4e6a-a91c-80022caa4daf": persistentvolumes "pvc-2355cacb-b2da-4e6a-a91c-80022caa4daf" is forbidden: User "system:serviceaccount:cephcsi-e2e-f0cfe643:nfs-csi-provisioner" cannot update resource "persistentvolumes" in API group "" at the cluster scope
I0720 12:02:57.004714       1 controller.go:1471] delete "pvc-2355cacb-b2da-4e6a-a91c-80022caa4daf": started
I0720 12:02:57.010197       1 connection.go:183] GRPC call: /csi.v1.Controller/DeleteVolume
I0720 12:02:57.010230       1 connection.go:184] GRPC request: {"secrets":"***stripped***","volume_id":"0001-0024-6be59c91-cd16-4814-811e-300690c167f3-0000000000000001-aed2ab46-0823-11ed-8c24-5e0233e9e911"}
I0720 12:02:57.014871       1 connection.go:186] GRPC response: {}
I0720 12:02:57.014934       1 connection.go:187] GRPC error: rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory
E0720 12:02:57.014992       1 controller.go:1481] delete "pvc-2355cacb-b2da-4e6a-a91c-80022caa4daf": volume deletion failed: rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory
W0720 12:02:57.015065       1 controller.go:989] Retrying syncing volume "pvc-2355cacb-b2da-4e6a-a91c-80022caa4daf", failure 1
E0720 12:02:57.015105       1 controller.go:1007] error syncing volume "pvc-2355cacb-b2da-4e6a-a91c-80022caa4daf": rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory
```

nixpanic commented:
The finalizer is removed from the PV with `Update()` here:

https://github.com/kubernetes-csi/external-provisioner/blob/b67655857d4800483928fe62e17660175bcb6ef0/vendor/sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go#L1519

nixpanic commented:
The change in the RBAC helps! Unfortunately the csi-external-provisioner uses `update` and not `patch`. So this happens now too (probably a race condition):
```
I0720 15:12:42.123725       1 controller.go:1471] delete "pvc-633d1c1f-55c3-4110-b8c6-394ec1f87092": started
I0720 15:12:42.126382       1 connection.go:183] GRPC call: /csi.v1.Controller/DeleteVolume
I0720 15:12:42.126406       1 connection.go:184] GRPC request: {"secrets":"***stripped***","volume_id":"0001-0024-62806a91-2f38-465c-8768-7afde251ab88-0000000000000001-0e6567c0-083e-11ed-8c44-4e32a48eec16"}
I0720 15:12:42.259193       1 connection.go:186] GRPC response: {}
I0720 15:12:42.259249       1 connection.go:187] GRPC error: <nil>
I0720 15:12:42.259264       1 controller.go:1486] delete "pvc-633d1c1f-55c3-4110-b8c6-394ec1f87092": volume deleted
I0720 15:12:42.273242       1 controller.go:1523] delete "pvc-633d1c1f-55c3-4110-b8c6-394ec1f87092": failed to remove finalizer for persistentvolume: Operation cannot be fulfilled on persistentvolumes "pvc-633d1c1f-55c3-4110-b8c6-394ec1f87092": the object has been modified; please apply your changes to the latest version and try again
W0720 15:12:42.273284       1 controller.go:989] Retrying syncing volume "pvc-633d1c1f-55c3-4110-b8c6-394ec1f87092", failure 0
E0720 15:12:42.273339       1 controller.go:1007] error syncing volume "pvc-633d1c1f-55c3-4110-b8c6-394ec1f87092": Operation cannot be fulfilled on persistentvolumes "pvc-633d1c1f-55c3-4110-b8c6-394ec1f87092": the object has been modified; please apply your changes to the latest version and try again
I0720 15:12:42.273365       1 controller.go:1471] delete "pvc-633d1c1f-55c3-4110-b8c6-394ec1f87092": started
```

A 2nd delete fails differently, because the NFS-provisioner already deleted the volume, and the `DeleteVolume` is not completely idempotent and returns an error:
```
I0720 15:12:42.275684       1 connection.go:183] GRPC call: /csi.v1.Controller/DeleteVolume
I0720 15:12:42.275755       1 connection.go:184] GRPC request: {"secrets":"***stripped***","volume_id":"0001-0024-62806a91-2f38-465c-8768-7afde251ab88-0000000000000001-0e6567c0-083e-11ed-8c44-4e32a48eec16"}
I0720 15:12:42.280482       1 connection.go:186] GRPC response: {}
I0720 15:12:42.280526       1 connection.go:187] GRPC error: rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory
```

I'm going to add a 2nd commit to this PR that hopefully improves the DeleteVolume behaviour.

yati1998 commented:
@nixpanic I see the golangci-lint to fail with gofmt error.
```
internal/nfs/controller/errors.go:27: File is not `gofmt`-ed with `-s` (gofmt)
	ErrNotFound           = errors.New("not found")
	ErrExportNotFound     = fmt.Errorf("NFS-export %w", ErrNotFound)```

nixpanic commented:
> @nixpanic I see the golangci-lint to fail with gofmt error.
> 
> ```
> internal/nfs/controller/errors.go:27: File is not `gofmt`-ed with `-s` (gofmt)
> 	ErrNotFound           = errors.New("not found")
> 	ErrExportNotFound     = fmt.Errorf("NFS-export %w", ErrNotFound)```
> ```

Ah, yes, need to check that, and also make sure to document the errors. This is just a try to see if it works with this correction, or if more is needed.

yati1998 commented:
> > @nixpanic I see the golangci-lint to fail with gofmt error.
> > ```
> > internal/nfs/controller/errors.go:27: File is not `gofmt`-ed with `-s` (gofmt)
> > 	ErrNotFound           = errors.New("not found")
> > 	ErrExportNotFound     = fmt.Errorf("NFS-export %w", ErrNotFound)```
> > ```
> 
> Ah, yes, need to check that, and also make sure to document the errors. This is just a try to see if it works with this correction, or if more is needed.

yeah sure, Thnks for the commit @nixpanic

nixpanic committed:
e2e: correct failure logging for NFS

Some of the steps still refer to CephFS, likely missed some replacements
while copy/pasting. The logging is a little confusing when messages
claim something with CephFS failed, but the test is about NFS.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

Rakshith-R commented on internal/nfs/controller/volume.go:
```diff
@@ -251,18 +253,18 @@ func (nv *NFSVolume) deleteExportCommand(cmd, nfsCluster string) []string {
 	fsName, err := fs.GetFsName(nv.ctx, nv.fscID)
 	if err != nil {
-		return "", fmt.Errorf("failed to get filesystem name for ID %x: %w", nv.fscID, err)
+		return "", fmt.Errorf("filesystem name for ID %x %w: %v", nv.fscID, ErrNotFound, err)
```
What if there is some other error while trying to fetch fsname or metadatapool. 
This change assumes, any error to be ErrNotFound and continues deletion. 
There maybe chances of intermittent connection errors or something else.

nixpanic replied:
Yes, but I do not think that these errors can be checked easily. Do you prefer to handle the error for the filesystem-id as a generic error?

Rakshith-R replied:
> Yes, but I do not think that these errors can be checked easily. Do you prefer to handle the error for the filesystem-id as a generic error?

Any error other than fs/metadatapool not found can lead to stale nfs exports in this case, right ? 
Maybe check for 'not found' in the actual error msg, only then add this `ErrNotFound` so it can be checked in higher levels ?

Madhu-1 replied:
fs.GetFsName already returning util.ErrPoolNotFound error if the fs doesn't exist just use it?

Madhu-1 committed:
ci: update mergify rules for kubernetes 1.24

Updating mergify rules to consider CI run on
Kubernetes 1.24 and discard CI run on kubernetes
1.21 as we no longer need it.

updates: #3086

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
e2e: remove kubernetes 1.22 check

We run CI jobs on kubernetes 1.22 by default
and we dont need to have a check to make sure
we have atleast Kubernetes 1.22 for few tests.
As we have CI runs on 1.22 by default, Removing
unwanted check.

updates: #3086
depends-on #3255

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
e2e: add deadcode nolint for k8sVersionGreaterEquals

k8sVersionGreaterEquals is not used anywhere but it
will be used in future if we need to have a kubernetes
version check. adding nolint for it now to skip it
from static check.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

yati1998 commented:
Apology for the mess, not sure what went wrong. Closing this PR..Here's the new one https://github.com/ceph/ceph-csi/pull/3265 . I will address these comments there.
